sw/device/lib/dif/dif_flash_ctrl.c

```diff
@@ -0,0 +1,1043 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 1.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_flash_ctrl.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+
+#include "flash_ctrl_regs.h"
+
+// Helper macros for statically checking register layout assumptions.
+#define EXPAND_REGION(id, name, suffix) \
+  FLASH_CTRL_MP_REGION_CFG##id##_##name##id##suffix
+
+#define ASSERT_REGION(name, suffix, id1, id2)                               \
+  _Static_assert(                                                           \
+      EXPAND_REGION(id1, name, suffix) == EXPAND_REGION(id2, name, suffix), \
+      "Region layout changed.")
+
+#define ASSERT_REGION_UNIFORMITY(name, suffix) \
+  ASSERT_REGION(name, suffix, 0, 1);           \
+  ASSERT_REGION(name, suffix, 1, 2);           \
+  ASSERT_REGION(name, suffix, 2, 3);           \
+  ASSERT_REGION(name, suffix, 3, 4);           \
+  ASSERT_REGION(name, suffix, 4, 5);           \
+  ASSERT_REGION(name, suffix, 5, 6);           \
+  ASSERT_REGION(name, suffix, 6, 7)
+
+// Helper macro to check that region registers are contiguous.
+#define ASSERT_REGION_CONTINUITY(id1, id2)                       \
+  _Static_assert(FLASH_CTRL_MP_REGION_CFG##id1##_REG_OFFSET +    \
+                         FLASH_CTRL_PARAM_REG_WIDTH / 8 ==       \
+                     FLASH_CTRL_MP_REGION_CFG##id2##_REG_OFFSET, \
+                 "Region register layout is not contiguous.")
+
+// Helper macros only check the first 8 regions and must be updated if regions
+// are added.
+_Static_assert(FLASH_CTRL_PARAM_NUMREGIONS == 8,
+               "Total number of regions changed.");
+
+// Static assert all regions are laid out the same.
+// TODO Remove this when MP region common parameters are added.
+ASSERT_REGION_UNIFORMITY(EN, );
+ASSERT_REGION_UNIFORMITY(RD_EN, );
+ASSERT_REGION_UNIFORMITY(PROG_EN, );
+ASSERT_REGION_UNIFORMITY(ERASE_EN, );
+ASSERT_REGION_UNIFORMITY(BASE, _MASK);
+ASSERT_REGION_UNIFORMITY(BASE, _OFFSET);
+ASSERT_REGION_UNIFORMITY(SIZE, _MASK);
+ASSERT_REGION_UNIFORMITY(SIZE, _OFFSET);
+ASSERT_REGION_UNIFORMITY(PARTITION, );
+
+// Static assert register layout is contiguous.
+ASSERT_REGION_CONTINUITY(0, 1);
+ASSERT_REGION_CONTINUITY(1, 2);
+ASSERT_REGION_CONTINUITY(2, 3);
+ASSERT_REGION_CONTINUITY(3, 4);
+ASSERT_REGION_CONTINUITY(4, 5);
+ASSERT_REGION_CONTINUITY(5, 6);
+ASSERT_REGION_CONTINUITY(6, 7);
+
+const static bitfield_field32_t CONTROL_OP_BITFIELD = {
```

Comment:
+1 to what the guys said regarding the `kConst`, but I like the idea of creating constants for the fields that are used more than once.